language-server/src/test/java/org/sonarlint/languageserver/ServerMainTest.java

```diff
@@ -0,0 +1,308 @@
+  }
+
+  @Test
+  public void analyzeSimpleJsFileOnSave() throws Exception {
+    File tempFile = temp.newFile("foo.js");
+    aut.getTextDocumentService()
+      .didSave(new DidSaveTextDocumentParams(new TextDocumentIdentifier(tempFile.toURI().toString()), "function foo() {\n  alert('toto');\n}"));
+
+    waitForDiagnostics();
+
+    assertThat(client.diagnosticsParamsList.get(0).getDiagnostics())
+      .extracting("range.start.line", "range.start.character", "range.end.line", "range.end.character", "code", "source", "message", "severity")
+      .containsExactly(tuple(1, 2, 1, 15, "javascript:S1442", "sonarlint", "Remove this usage of alert(...). (javascript:S1442)", DiagnosticSeverity.Information));
+  }
+
+  @Test
+  public void cleanDiagnosticsOnClose() throws Exception {
+    File tempFile = temp.newFile("foo.js");
+    aut.getTextDocumentService()
+      .didClose(new DidCloseTextDocumentParams(new TextDocumentIdentifier(tempFile.toURI().toString())));
+
+    waitForDiagnostics();
+
+    assertThat(client.diagnosticsParamsList.get(0).getDiagnostics()).isEmpty();
+  }
+
+  @Test
+  public void testCodeActionRuleDescription() throws Exception {
+    File tempFile = temp.newFile("foo.js");
+    VersionedTextDocumentIdentifier docId = new VersionedTextDocumentIdentifier(1);
+    docId.setUri(tempFile.toURI().toString());
+    aut.getTextDocumentService()
+      .didChange(new DidChangeTextDocumentParams(docId, Arrays.asList(new TextDocumentContentChangeEvent("function foo() {\n  alert('toto');\n}"))));
+
+    waitForDiagnostics();
+
+    Diagnostic diagnostic = client.diagnosticsParamsList.get(0).getDiagnostics().get(0);
+
+    List<? extends Command> codeActions = aut.getTextDocumentService()
+      .codeAction(new CodeActionParams(new TextDocumentIdentifier(tempFile.toURI().toString()), new Range(new Position(1, 4), new Position(1, 4)),
+        new CodeActionContext(Arrays.asList(diagnostic))))
+      .get();
+
+    assertThat(codeActions).hasSize(1);
+
+    String ruleKey = (String) codeActions.get(0).getArguments().get(0);
+    int port = ((Double) codeActions.get(0).getArguments().get(1)).intValue();
+    assertThat(ruleKey).isEqualTo("javascript:S1442");
+
+    try {
+      IOUtils.toString(new URL("http://localhost:" + port + "/"));
+      Assert.fail("Expected exception");
+    } catch (Exception e) {
+      assertThat(e.getMessage()).contains("400");
+    }
+
+    try {
+      IOUtils.toString(new URL("http://localhost:" + port + "/?ruleKey=foo"));
+      Assert.fail("Expected exception");
+    } catch (Exception e) {
+      assertThat(e).isInstanceOf(FileNotFoundException.class);
+    }
+
+    String ruleDesc = IOUtils.toString(new URL("http://localhost:" + port + "/?ruleKey=" + ruleKey));
+    assertThat(ruleDesc).contains("\"alert(...)\" should not be used");
+  }
+
+  private void waitForDiagnostics() throws InterruptedException {
+    int maxLoop = 20;
+    do {
+      Thread.sleep(100);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this use of "Thread.sleep()". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2925)